Remotion/ObjectBinding/Web/UI/Controls/Validation/BocListValidationFailureRepository.cs

```diff
@@ -0,0 +1,315 @@
+      {
+        _rowFailures ??= new List<HandleableFailure>();
+        _rowFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public void AddCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition, IEnumerable<BusinessObjectValidationFailure> failures)
+      {
+        List<HandleableFailure> cellFailures;
+        _cellFailures ??= new Dictionary<IBocColumnDefinitionWithValidationSupport, List<HandleableFailure>>();
+        if (_cellFailures.TryGetValue(cellDefinition, out var existingFailureList))
+        {
+          cellFailures = existingFailureList;
+        }
+        else
+        {
+          cellFailures = new List<HandleableFailure>();
+          _cellFailures[cellDefinition] = cellFailures;
+        }
+
+        cellFailures.AddRange(failures.Select(e => new HandleableFailure(e)));
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetRowAndCellFailures ()
+      {
+        var failures = new List<HandleableFailure>();
+        failures.AddRange(GetRowFailures());
+        failures.AddRange(GetAllCellFailuresForRow());
+        return failures;
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetRowFailures ()
+      {
+        if (_rowFailures == null)
+          return Array.Empty<HandleableFailure>();
+
+        return _rowFailures.Where(e => !e.Handled).ToList();
+      }
+
+      private IReadOnlyCollection<HandleableFailure> GetAllCellFailuresForRow ()
+      {
+        if (_cellFailures == null)
+          return Array.Empty<HandleableFailure>();
+
+        return _cellFailures.SelectMany(c => c.Value.Where(e => !e.Handled)).ToList();
+      }
+
+      public IReadOnlyCollection<HandleableFailure> GetCellFailures (IBocColumnDefinitionWithValidationSupport cellDefinition)
+      {
+        if (_cellFailures == null)
+          return Array.Empty<HandleableFailure>();
+
+        if (!_cellFailures.TryGetValue(cellDefinition, out var handleableFailures))
+          return Array.Empty<HandleableFailure>();
+
+        return handleableFailures.Where(e => !e.Handled).ToList();
+      }
+
+      public bool HasValidationFailures ()
+      {
+        return _rowFailures is { Count: > 0 } || _cellFailures != null && _cellFailures.Values.SelectMany(e => e).Any();
+      }
+    }
+
+    private readonly List<HandleableFailure> _listValidationFailures;
+    private readonly Dictionary<IBusinessObject, RowFailureContainer> _rowValidationFailures;
+
+    public BocListValidationFailureRepository ()
+    {
+      _listValidationFailures = new List<HandleableFailure>();
+      _rowValidationFailures = new Dictionary<IBusinessObject, RowFailureContainer>();
+    }
+
+    /// <summary>
+    /// Adds <see cref="BusinessObjectValidationFailure"/>s that belong to the <see cref="BocList"/> as a whole.
+    /// </summary>
+    public void AddListValidationFailures (IEnumerable<BusinessObjectValidationFailure> validationFailures)
```

Comment:
Tip / please do: since we have an interface, you don't need to add method doc on the implementation. if you do want to make sure there's docs, you can use <inhertdoc />. Of course, if there's important add-ons that are needed, things are different. for the most part, that's not the case :)